src/app/CommandSender.cpp

```diff
@@ -61,10 +62,18 @@ CHIP_ERROR CommandSender::SendCommandRequest(NodeId aNodeId, Transport::AdminId
     VerifyOrExit(mpExchangeCtx != nullptr, err = CHIP_ERROR_NO_MEMORY);
     mpExchangeCtx->SetResponseTimeout(kImMessageTimeoutMsec);
 
-    err = mpExchangeCtx->SendMessage(
-        Protocols::InteractionModel::MsgType::InvokeCommandRequest, std::move(mCommandMessageBuf),
-        Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse).Set(Messaging::SendMessageFlags::kNoAutoRequestAck));
-    SuccessOrExit(err);
+    event.Type                              = DeviceLayer::DeviceEventType::kInteractionModelCommandRequest;
+    event.InteractionModelEvent.ExchangeCtx = mpExchangeCtx;
```

Comment:
This looks like we are allocating the exchange context on this thread but then we are posting the event ... to a different thread?  Is that the idea?  Which thread are we posting to?  Is it the thread that incoming messages get handled on?  Because if it _is_ then it seems like we would be allocating exchanges on that thread too (for unsolicited messages), and if so we are touching the exchange manager on both threads, and that is probably not going to work great at all.

As in, my gut feeling is that CommandSender should be running on the same thread as message processing, so any asynchronicity needs to be above this level.

Reply:
The event is always posted to the chip event loop for processing, that is the same thread that incoming messages get handled on.

The exchange is allocated from the sending thread in this PR, and its handle get passed to the chip event loop within the event message. 

This PR is not complete, it just show one of approaches to solve the current racing issue by putting all essential  operations into chip main thread. We should define those essential operation as the CHIP service APIs. We should use dedicated lock to protect the preliminary APIs which could be called from multiple threads,  such as NewContext and NewWithData.

If you move posting event even further and make CommandSender driven by the event, then we don't need to protect NewContext.